torch_xla/csrc/dynamic_shape_detector.cpp

```diff
@@ -0,0 +1,254 @@
+#include "torch_xla/csrc/dynamic_shape_detector.h"
+
+#include <sstream>
+
+#include "torch_xla/csrc/runtime/debug_macros.h"
+
+namespace torch_xla {
+
+// Maximum number of allowed traces per function (i.e. session).
+static std::size_t max_allowed_traces_per_function = 1;
+
+TrieNode::TrieNode(absl::Span<const TrieValue> common_sequence,
+                   bool is_trace_boundary)
+    : common_sequence_(common_sequence.begin(), common_sequence.end()),
+      is_trace_boundary_(is_trace_boundary) {}
+
+bool TrieNode::IsLeaf() const { return children_.empty(); }
+
+void TrieNode::NewTraceNotAllowedError(std::optional<TrieValue> value,
+                                       std::size_t matched) {
+  std::ostringstream ostr;
+  ostr << "Maximum number of different traces allowed per function exceeded: "
+       << max_allowed_traces_per_function << std::endl;
+
+  if (value.has_value()) {
+    ostr << "Got: " << value->str << std::endl;
+  } else {
+    ostr << "Reached the end of the function at: "
+         << common_sequence_[matched - 1].str << std::endl;
+  }
+
+  if (common_sequence_.size() > matched) {
+    ostr << "Expected: " << common_sequence_[matched].str << std::endl;
+  } else {
+    ostr << "Expected either of:" << std::endl;
+    for (auto& pair : children_) {
+      ostr << "  - " << pair.second->common_sequence_.front().str << std::endl;
+    }
+  }
+
```

Comment:
I think we should also dump the current python trace, which would help people to figure out where to fix. Check https://github.com/pytorch/xla/blob/master/torch_xla/csrc/debug_util.cpp#L125-L131 for the example.

Reply:
Here's an example running `test_trace_limit_exceeded_common_sequence_mismatch`:

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/usr/local/lib/python3.10/unittest/case.py", line 591, in run
    self._callTestMethod(testMethod)
  File "/usr/local/lib/python3.10/unittest/case.py", line 549, in _callTestMethod
    method()
  File "xla/test/test_dynamic_shapes_detector.py", line 104, in test_trace_limit_exceeded_common_sequence_mismatch
    self._run_and_compare(foo, args=(inp, 2), allowed_traces=allowed_traces)
  File "xla/test/test_dynamic_shapes_detector.py", line 20, in _run_and_compare
    optout = optf(*args)
  File "/usr/local/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "xla/test/test_dynamic_shapes_detector.py", line 93, in foo
    return x * 5
RuntimeError: torch_xla/csrc/dynamic_shape_detector.cpp:41 : Maximum number of different traces allowed per function exceeded: 1
Got: [] aten::mul, xla_shape=f32[10]{0}, dynamic_dims: ()
Expected: [] aten::add, xla_shape=f32[10]{0}, dynamic_dims: ()
```

We already have the python trace, since this check is being done incrementally, every time a new IR node is created.

Reply:
Ok this is nice, I guess we can go a step further where for every TRIE node we also store the python stack trace for the current node. This way when we raise the runtime error, we can also show that here is the python stack we expects, but now we hit here. It would be easier to debug this way.

we can implement this as a follow up through.